services/requester/pool.go

```diff
@@ -0,0 +1,124 @@
+package requester
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/onflow/flow-go-sdk"
+	"github.com/onflow/flow-go/fvm/evm/types"
+	gethTypes "github.com/onflow/go-ethereum/core/types"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-evm-gateway/models"
+)
+
+const evmErrorRegex = `evm_error=(\d+)`
+
+// todo this is a simple implementation of the transaction pool that is mostly used
+// to track the status of submitted transaction, but transactions will always be submitted
+// right away, future improvements can make it so the transactions are collected in the pool
+// and after submitted based on different strategies.
+
+type TxPool struct {
+	logger zerolog.Logger
+	client *CrossSporkClient
+	pool   *sync.Map
+	// todo add a broadcaster for pending transaction streaming
+	// todo add methods to inspect transaction pool state
+}
+
+func NewTxPool(client *CrossSporkClient, logger zerolog.Logger) *TxPool {
+	return &TxPool{
+		logger: logger.With().Str("component", "tx-pool").Logger(),
+		client: client,
+		pool:   &sync.Map{},
+	}
+}
+
+// Send flow transaction that executes EVM run function which takes in the encoded EVM transaction.
+// The flow transaction status is awaited and an error is returned in case of a failure in submission,
+// or an EVM validation error.
+// Until the flow transaction is sealed the transaction will stay in the transaction pool marked as pending.
+func (t *TxPool) Send(
+	ctx context.Context,
+	flowTx *flow.Transaction,
+	evmTx *gethTypes.Transaction,
+) error {
+	if err := t.client.SendTransaction(ctx, *flowTx); err != nil {
+		return err
+	}
+
+	// add to pool
+	t.pool.Store(evmTx.Hash(), evmTx)
+
+	const fetchInterval = time.Millisecond * 500
+	const fetchTimeout = time.Minute * 3
+	ticker := time.NewTicker(fetchInterval)
+	timeout := time.NewTimer(fetchTimeout)
+
+	defer func() {
+		t.pool.Delete(evmTx.Hash())
+		timeout.Stop()
+		ticker.Stop()
+	}()
+
+	for {
+		select {
+		case <-ticker.C:
+			res, err := t.client.GetTransactionResult(ctx, flowTx.ID())
```

Comment:
That's very nice :clap: This will allow us to better handle the keys from the rotation signers.